src/Middleware.php

```diff
@@ -3,31 +3,138 @@
+     * Handle the incoming request.
+     *
+     * @param  \Illuminate\Http\Request  $request
+     * @param  Closure  $next
+     * @return Response
+     */
+    public function handle(Request $request, Closure $next)
+    {
+        Inertia::share($this->share($request));
+
+        Inertia::version(function () use ($request) {
+            return $this->version($request);
+        });
+
+        $response = $next($request);
+        $response = $this->checkVersion($request, $response);
+        $response = $this->changeRedirectCode($request, $response);
+
+        return $response;
+    }
 
-        if ($request->method() === 'GET' && $request->header('X-Inertia-Version', '') !== Inertia::getVersion()) {
+    /**
+     * In the event that the asset version changes, Inertia will automatically reload
+     * the page in order to ensure the application keeps working for the end user.
+     *
+     * @param  Request  $request
+     * @param  Response  $response
+     * @return Response
+     */
+    public function checkVersion(Request $request, Response $response)
+    {
+        if ($request->header('X-Inertia') &&
+            $request->method() === 'GET' &&
+            $request->header('X-Inertia-Version', '') !== Inertia::getVersion()
+        ) {
             if ($request->hasSession()) {
                 $request->session()->reflash();
             }
 
-            return Response::make('', 409, ['X-Inertia-Location' => $request->fullUrl()]);
+            return Inertia::location($request->fullUrl());
         }
 
-        if ($response instanceof Redirect && $response->getStatusCode() === 302 && in_array($request->method(), ['PUT', 'PATCH', 'DELETE'])) {
+        return $response;
+    }
+
+    /**
+     * Changes the status code during an Inertia redirect, so that the browser
+     * makes the request as GET instead, avoiding unsupported method calls.
+     *
+     * @param  Request  $request
+     * @param  Response  $response
+     * @return Response
+     */
+    public function changeRedirectCode(Request $request, Response $response)
+    {
+        if ($response instanceof RedirectResponse &&
+            $request->header('X-Inertia') &&
+            $response->getStatusCode() === 302 &&
+            in_array($request->method(), ['PUT', 'PATCH', 'DELETE'])
+        ) {
             $response->setStatusCode(303);
         }
 
         return $response;
     }
+
+    /**
+     * Resolves and prepares validation errors in such a way
+     * that they are easier to use in the Inertia client.
+     *
+     * @param  Request  $request
+     * @return object
+     */
+    public function resolveValidationErrors(Request $request)
+    {
+        if (! $request->session()->has('errors')) {
+            return (object) [];
+        }
+
+        return (object) Collection::make($request->session()->get('errors')->getBags())->map(function ($bag) {
+            return (object) Collection::make($bag->messages())->map(function ($errors) {
```

Comment:
We're already using both the `public_path()` and `config()` helpers, might as well use the `collect()` helper as well. 😆

```suggestion
        return (object) collect($request->session()->get('errors')->getBags())->map(function ($bag) {
            return (object) collect($bag->messages())->map(function ($errors) {
```